titan-compiler/checker.lua

```diff
@@ -689,43 +621,155 @@ local function checkfunc(node, st, errors)
     end
 end
 
--- Third typechecking pass over the module, checks function bodies
+-- Checks function bodies
 --   ast: AST for the whole module
 --   st: symbol table
 --   errors: list of compile-time errors
-local function thirdpass(ast, st, errors)
-    for _, tlnode in ipairs(ast) do
-        if not tlnode._ignore then
-            local tag = tlnode._tag
-            if tag == "TopLevel_Func" then
-                st:with_block(checkfunc, tlnode, st, errors)
+local function checkbodies(ast, st, errors)
+    for _, node in ipairs(ast) do
+        if not node._ignore then
+            if node._tag == "TopLevel_Func" then
+                st:with_block(checkfunc, node, st, errors)
             end
         end
     end
 end
 
--- Gets type information for all imported modules and puts
--- it in the symbol table
-local function importpass(ast, st, errors, loader)
-    for _, tlnode in ipairs(ast) do
-        if tlnode._tag == "TopLevel_Import" then
-            local name = tlnode.modname
-            if st:find_dup(tlnode.localname) then
-                typeerror(errors, "duplicate declaration for " .. tlnode.localname, tlnode._pos)
-                tlnode._ignore = true
-            else
-                local modtype, errs = checker.checkimport(name, loader)
-                if modtype then
-                    tlnode._type = modtype
-                    for _, err in ipairs(errs) do
-                        table.insert(errors, err)
-                    end
-                    st:add_symbol(tlnode.localname, tlnode)
-                else
-                    tlnode._type = types.Nil
-                    typeerror(errors, "problem loading module '%s': %s", tlnode._pos, name, errs)
+-- Verify if an expression is constant
+local function isconst(node)
+    local tag = node._tag
+    if tag == "Exp_Nil" or
+       tag == "Exp_Bool" or
+       tag == "Exp_Integer" or
+       tag == "Exp_Float" or
+       tag == "Exp_String" then
+        return true
+
+    elseif tag == "Exp_InitList" then
+        local const = true
+        for _, field in ipairs(node.fields) do
+            const = const and isconst(field.exp)
+        end
+        return const
+
+    elseif tag == "Exp_Call" then
+        return false
+
+    elseif tag == "Exp_Var" then
+        if node.var._tag == "Var_Name" then
+            return true
```

Comment:
For now, I think we should consider these as not constants. The code generator currently evaluates the top-level in top down order so if we let it try to evaluate variables defined further down it is going to try to read an uninitialized variable.

We might be able to allow variables in a future version but in addition to changing the code generator we would also need to be careful about forbidding nonsensical mutual declarations such as 

    local a = b
    local b = a